### PR TITLE
Add support for Golang on Linux (amd64)

### LINKED ISSUE #225: Support Go on Linux
Go on Linux doesn't use libc and calls syscalls directly. To support it, we plan on hooking Go's syscalls functions and taking it from there.

### PR DESCRIPTION
**Pull request by infiniteregrets**
This PR adds support for Golang hooks. I will update here with more info once the draft is finished.
Closes #225

### CODE REVIEW

File: mirrord-cli/src/config.rs [hunk 1]
```diff
@@ -81,4 +81,7 @@ pub(super) struct ExecArgs {
     /// Use an Ephemeral Container to mirror traffic.
     #[clap(short, long, value_parser)]
     pub ephemeral_container: bool,
+
+    #[clap(short, long, value_parser)]
+    pub enable_go_hooks: bool,
```
Comment (aviramha) [inline mirrord-cli/src/config.rs:86]: I don't think it needs to be a flag..

File: mirrord-cli/src/main.rs [hunk 2]
```diff
@@ -75,7 +75,7 @@ fn exec(args: &ExecArgs) -> Result<()> {
         "Launching {:?} with arguments {:?}",
         args.binary, args.binary_args
     );
-
+    std::env::set_var("MIRRORD_DEBUG_BINARY", args.binary.clone());
```
Comment (aviramha) [inline mirrord-cli/src/main.rs:78]: ?

File: mirrord-layer/src/go/hooks.rs [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+            param2 as *const libc::sockaddr,
+            param3 as u32,
+        ) as i64, //TODO: check if this argument passing is right?
+        libc::SYS_listen => listen_detour(param1 as i32, param2 as i32) as i64,
+        libc::SYS_accept4 => accept_detour(
+            param1 as i32,
+            param2 as *mut libc::sockaddr,
+            param3 as *mut u32,
+        ) as i64,
+        //10000 => 1, // Ask Aviram what is this for?
```
Comment (aviramha) [inline mirrord-layer/src/go/hooks.rs:111]: this was for testing, you can remove it

**Review by aviramha — CHANGES_REQUESTED** [review]

File: mirrord-layer/src/go/hooks.rs [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        "mov    rdi, rsi",
+        "mov    rsi, rdx",
+        "mov    rdx, rcx",
+        "syscall",
+        "ret",
+        options(noreturn)
+    )
+}
+
+pub(crate) fn enable_socket_hooks(interceptor: &mut Interceptor, binary: &str) {
```
[the remark this answers is not in the document]
Comment (infiniteregrets) [reply mirrord-layer/src/go/hooks.rs:134]: yeah we do, this is not complete. I was looking into a way to match symbols when the binary is run with mirrord

File: CHANGELOG.md [hunk 5]
```diff
@@ -10,6 +10,7 @@ Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how
 
 ### Added
 - Add a flag for the agent, `--ephemeral-container`, to correctly refer to the filesystem i.e. refer to root path as `/proc/1/root` when the flag is on, otherwise `/`.
+- Add support for Golang (Gin framework) on amd64 (x86-64).
```
Comment (aviramha) [inline CHANGELOG.md:13]: I don't think this is `Gin` specific?

File: mirrord-layer/src/go/hooks.rs [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+            "cmp    rax, -0xfff",
+            "jbe    3f",
+            "mov    QWORD PTR [rsp+0x28], -0x1",
+            "mov    QWORD PTR [rsp+0x30], 0x0",
+            "neg    rax",
+            "mov    QWORD PTR [rsp+0x38], rax",
+            "xorps  xmm15, xmm15",
+            "mov    r14, qword ptr FS:[0xfffffff8]",
+            "ret",
+            // Not exactly sure why were are doing this here..?
```
Comment (aviramha) [inline mirrord-layer/src/go/hooks.rs:93]: The lines above or below?

File: mirrord-layer/src/go/hooks.rs [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+            "mov    r8, r9",
+            "mov    r9, qword ptr[rsp]",
+            "syscall",
+            "ret",
+            options(noreturn)
+        )
+    }
+
+    pub(crate) fn enable_socket_hooks(interceptor: &mut Interceptor, binary: &str) {
+        /*
```
Comment (aviramha) [inline mirrord-layer/src/go/hooks.rs:387]: imo put it in the fn docstring

File: mirrord-layer/src/go/hooks.rs [hunk 8]
```diff
@@ -0,0 +1,423 @@
+#[cfg(target_os = "linux")]
```
Comment (aviramha) [inline mirrord-layer/src/go/hooks.rs:1]: don't make it a separate directory just for one file.

File: mirrord-layer/src/go/hooks.rs [hunk 9]
```diff
@@ -0,0 +1,423 @@
...
+    }
+
+    /// [Naked function] maps to gasave_systemstack_switch, called by asmcgocall.abi0
+    #[no_mangle]
+    #[naked]
+    unsafe extern "C" fn go_systemstack_switch() {
+        asm!(
+            // 0xdd9 - is the value taken from objdump, ghidra maps it to
+            // `runtime.systemstack_switch.abi0`, which just returns.
+            // TODO: fix the address so the symbol is loaded correctly
```
Comment (aviramha) [inline mirrord-layer/src/go/hooks.rs:276]: I don't think it's needed - if I remember correctly, the systemstack_switch exists just to be a marker

File: mirrord-layer/src/go/hooks.rs [hunk 10]
```diff
@@ -0,0 +1,423 @@
...
+            libc::SYS_accept => accept_detour(param1 as _, param2 as _, param3 as _) as i64,
+            libc::SYS_close => close_detour(param1 as _) as i64,
+            _ => syscall_3(syscall, param1, param2, param3),
+        };
+        debug!(
+            "c_abi_syscall_handler >> result -> {res:?}, errorno -> {:?}",
+            errno().0
+        );
+        match res {
+            -1 => -errno().0 as i64,
```
Comment (aviramha) [inline mirrord-layer/src/go/hooks.rs:324]: this might be a valid return code of the syscall case?

File: mirrord-layer/src/go/hooks.rs [hunk 11]
```diff
@@ -0,0 +1,423 @@
...
+    }
+
+    /// [Naked function] maps to runtime.abort.abi0, called by `go_systemstack_switch`
+    #[no_mangle]
+    #[naked]
+    unsafe extern "C" fn go_runtime_abort() {
+        asm!("int 0x3", "jmp go_runtime_abort", options(noreturn));
+    }
+
+    /// [Naked function] Syscall & Rawsyscall handler - supports upto 4 params, used for socket,
```
Comment (aviramha) [inline mirrord-layer/src/go/hooks.rs:300]: it's not naked?

File: mirrord-layer/src/go/hooks.rs [hunk 12]
```diff
@@ -0,0 +1,423 @@
...
+            "c_abi_syscall_handler >> result -> {res:?}, errorno -> {:?}",
+            errno().0
+        );
+        match res {
+            -1 => -errno().0 as i64,
+            _ => res as i64,
+        }
+    }
+
+    /// [Naked function] Syscall & Syscall6 handler - supports upto 6 params, mainly used for
```
Comment (aviramha) [inline mirrord-layer/src/go/hooks.rs:329]: not naked?

File: mirrord-layer/src/go/hooks.rs [hunk 13]
```diff
@@ -0,0 +1,423 @@
...
+            libc::SYS_accept4 => {
+                accept4_detour(param1 as _, param2 as _, param3 as _, param4 as _) as i64
+            }
+            _ => syscall_6(syscall, param1, param2, param3, param4, param5, param6),
+        };
+        debug!(
+            "c_abi_syscall_handler >> result -> {res:?}, errorno -> {:?}",
+            errno().0
+        );
+        match res {
```
Comment (aviramha) [inline mirrord-layer/src/go/hooks.rs:352]: same as above

File: mirrord-layer/src/go/hooks.rs [hunk 14]
```diff
@@ -0,0 +1,423 @@
...
+            options(noreturn)
+        )
+    }
+
+    /// Note: We only hook "RawSyscall", "Syscall6", and "Syscall" because for our usecase,
+    /// when testing with "Gin", only these symbols were used to make syscalls.
+    /// Refer:
+    ///   - File zsyscall_linux_amd64.go generated using mksyscall.pl.
+    ///   - https://cs.opensource.google/go/go/+/refs/tags/go1.18.5:src/syscall/syscall_unix.go
+    pub(crate) fn enable_socket_hooks(interceptor: &mut Interceptor, binary: &str) {
```
Comment (aviramha) [inline mirrord-layer/src/go/hooks.rs:403]: I'm okay with not doing file hooks right now, but 99% we need to do similar for file, so please open an issue so we won't forget it :)

**Review by aviramha — CHANGES_REQUESTED** [review]

File: mirrord-layer/src/go_hooks.rs [hunk 15]
```diff
@@ -0,0 +1,429 @@
...
+            }
+        };
+        debug!(
+            "c_abi_syscall6_handler (detour) >> result -> {res:?}, errorno -> {:?}",
+            errno().0
+        );
+        match res {
+            -1 => -errno().0 as i64,
+            _ => res as i64,
+        }
```
Comment (meowjesty) [inline mirrord-layer/src/go_hooks.rs:327]: `errorno` typo?

File: mirrord-layer/src/go_hooks.rs [hunk 16]
```diff
@@ -0,0 +1,429 @@
+#[cfg(target_os = "linux")]
+#[cfg(target_arch = "x86_64")]
+pub(crate) mod go_socket_hooks {
```
Comment (aviramha) [inline mirrord-layer/src/go_hooks.rs:3]: I think the cfg is only needed in `lib.rs`, also probably don't need to mod here?

Comment (infiniteregrets) [reply mirrord-layer/src/go_hooks.rs:3]: if we don't do so, we get dead code warnings

Comment (aviramha) [reply mirrord-layer/src/go_hooks.rs:3]: ok

**Review by aviramha — APPROVED** [review]
image

**Pull request merged**